feat: check cids in denylist before providing

olizilla opened the pull request:
- cap the max number of cids we'll accept in a single message. 
  - We're seeing 20k spikes in our `bitswap-pending-entries` metrics per node, so I'm putting in a hard cap of 500 wanted cids per message that we'll process. The caller can ask again if they need more. This also means i can put a sensible cap on how many cids the denylist service should handle in a batch.
- check batches of cids against our denylist api.
- cache entries that are on the denylist; they are rarely removed.
- use cache to avoid asking about items we already know, and as a fallback if denylist service cannot be reached.
- add `bitswap-denied` counter metric to see how many CIDs we skip due to being on the denylist


see: batch endpoint for denylist api – https://github.com/web3-storage/reads/pull/166
see: set DENYLIST_URL in env - https://github.com/elastic-ipfs/bitswap-peer-deployment/pull/99

License: MIT

olizilla commented on src/deny.js:
```diff
@@ -0,0 +1,63 @@
+    if (res.ok) {
+      const denylist = new Set(await res.json())
+      for (const cidStr of denylist.values()) {
+        denylistCache.set(cidToKey(CID.parse(cidStr)), true)
```
these cids are never pulled out of the cache... so items that do get removed from a badbits list won't get updated here... but our bitswap peer nodes tend to get restarted every day, so i'm not going to do extra work here... if this PR magically makes our pods less crashy, then we can review this.

vasco-santos replied:
could this comment be also added in the code file? I feel it would be nice to have context there, and maybe an issue created

olizilla commented on package.json:
```diff
@@ -28,6 +28,8 @@
     "libp2p": "0.42.2",
     "lru-cache": "7.14.1",
     "mnemonist": "0.39.5",
+    "multiformats": "^10.0.3",
```
10 because this code already depends on it at that version but previously didn't declare that dep, so would use a random one of 3 different major versions that are in the dep tree. Version 11+ doesn't work with the test helpers due to https://github.com/multiformats/js-multiformats/blob/4a36fb7ee49edb4300267b90301ef0e4300cbc46/src/cid.js#L305

olizilla commented on src/deny.js:
```diff
@@ -0,0 +1,66 @@
+        denylistCache.set(cidToKey(CID.parse(cidStr)), true)
+      }
+      // all `cancel`s go through, and items not on the denylist
+      return entries.filter(entry => entry.cancel || !denylist.has(entry.cid.toString()))
```
using `get` here to update the last used time. In this LRU impl `has` tests do not update that.

olizilla replied:
ah, confusing. `denylist` here is the set of CIDs returned from the denylist api, that represents the CIDs that should not be served. Is a Set not a Map

elsewhere we have `denylistCache` which is an LRU, and requires us to store a value for a our CID key, and that has a Map-like api, but we must call .get to update the last used time. I will tweak the names to make clearer.

vasco-santos commented on src/config.js:
```diff
@@ -74,7 +74,9 @@ export function makeConfig () {
-    allowReadinessTweak: process.env.ALLOW_READINESS_TWEAK === 'true'
+    allowReadinessTweak: process.env.ALLOW_READINESS_TWEAK === 'true',
+
+    denylistUrl: process.env.DENYLIST_URL ? new URL(process.env.DENYLIST_URL) : undefined
```
should this be added to the `.env.sample`? 
We also need to get this set in the deployment

olizilla commented:
@vasco-santos @alanshaw this PR changed a little since you last looked so please take a look if you have a moment.

- added `bitswap-denied` counter metric to track how many CIDs we ignore due to being on the denylist
- made the logic in deny.js more readable, particularly around the nuances of how we use the cached info vs the current denylist we get back from the api
- simplified the cache keys to just use cid strings. trades off catching alt cid formulations at the cache layer for simplicity and not having to cid.parse every item in the response list to dig out the multihash.
